narwhal/worker/src/handlers.rs

```diff
@@ -90,16 +98,61 @@ pub struct PrimaryReceiverHandler {
 
 #[async_trait]
 impl PrimaryToWorker for PrimaryReceiverHandler {
-    async fn send_message(
+    async fn reconfigure(
         &self,
-        request: anemo::Request<PrimaryWorkerMessage>,
+        request: anemo::Request<WorkerReconfigureMessage>,
     ) -> Result<anemo::Response<()>, anemo::rpc::Status> {
-        let message = request.into_body();
+        let message = request.into_body().message;
+        match &message {
+            ReconfigureNotification::NewEpoch(new_committee) => {
+                self.committee.swap(Arc::new(new_committee.clone()));
+
+                // TODO: duplicated code in this file.
+                // Update the worker cache.
+                self.worker_cache.swap(Arc::new(WorkerCache {
+                            epoch: new_committee.epoch,
+                            workers: new_committee.keys().iter().map(|key|
+                                (
+                                    (*key).clone(),
+                                    self.worker_cache
+                                        .load()
+                                        .workers
+                                        .get(key)
+                                        .tap_none(||
+                                            warn!("Worker cache does not have a key for the new committee member"))
+                                        .unwrap_or(&WorkerIndex(BTreeMap::new()))
+                                        .clone()
+                                )).collect(),
+                        }));
```

Comment:
Could we just factor this out into a method on the PrimaryReceiverHandler so that it can be deduped?

Reply:
We could. I just cut & pasted this code wholesale into the handler from synchronizer for this PR, so I was trying to limit additional diffs with concomitant risk of introducing a bug. Do you think it's better to do now or in a follow up pr? (I could see arguments either way. I guess I am used to trying to separate code moves from refactoring, but Github tooling for displaying a move differently from a diff is not as good as what I was previously used to...)

Reply:
Per our huddle, will add a second commit for this to this pr